static/js/components/dashboard/courses/StatusMessages.js

```diff
@@ -0,0 +1,227 @@
+
+  if (firstRun.status === STATUS_PAID_BUT_NOT_ENROLLED) {
+    const contactHref = `mailto:${SETTINGS.support_email}`;
+    return S.Just([{
+      message: <div>
+        {"Something went wrong. You paid for this course but are not enrolled. "}
+        <a href={contactHref}>
+          Contact us for help.
+        </a>
+      </div>
+    }]);
+  }
+
+  // Course run isn't enrollable, user never enrolled
+  if (!isEnrollableRun(firstRun) && !R.any(userIsEnrolled, course.runs)) {
+    return S.Just([{
+      message: "There are no future course runs scheduled."
+    }]);
+  }
+
+  // Course is running, user has already paid, we show no messages
+  if (courseUpcomingOrCurrent(firstRun) && paid) {
+    return S.Nothing; 
+  }
+
+  let messages = [];
+
+
+  if (coupon && coupon.content_type === COUPON_CONTENT_TYPE_COURSE && coupon.object_id === course.id) {
+    messages.push({ message: makeCouponMessage(coupon)});
+  }
+
+  // handle other 'in-progress' cases
+  if (courseUpcomingOrCurrent(firstRun) || firstRun.status === STATUS_CAN_UPGRADE) {
+    if (hasPaidForAnyCourseRun(course)) {
+      messages.push({message: "You are auditing this course."});
+    } else {
+      messages.push({
+        message: "You are auditing. To get credit, you need to pay for the course.",
+        action: courseAction(firstRun, COURSE_ACTION_PAY)
+      });
+    }
+    return S.Just(messages);
+  }
+
+  // all cases where courseRun is not currently in progress
+  // first, all cases where the user has already passed the course
+  if (isPassedCourseRun(firstRun)) {
+    // paid statuses
+    if (hasPaidForAnyCourseRun(course)) {
+      // exam is required, user has not yet passed it
+      if (exams && ! passedExam) {
+        messages.push(S.maybe(
+          { message: "The edX course is complete, but you need to pass the final exam." },
+          () => ({
+            message: <span>
+              The edX course is complete, but you need to pass the final exam.
+              {" If you want to re-take the course you can "}
+              <a onClick={() => setShowExpandedCourseStatus(course.id)}>
+                re-enroll.
+              </a>
+            </span>
+          }), futureEnrollableRun(course))
+        );
+
+        // this is the expanded message, which we should if the user clicks the link above
+        if (expandedStatuses.has(course.id)) {
+          messages.push(S.maybe(null, run => ({
+            message: `Next course starts ${formatDate(run.course_start_date)}.`,
+            action: courseAction(run, COURSE_ACTION_REENROLL),
+          }), futureEnrollableRun(course)));
+        }
+
+      } else {
+        messages.push({
+          // TODO add links
+          message: "You passed this course."
+        });
+      }
+      return S.Just(messages);
+    } else {
+      // user missed the payment due date
+      if (firstRun.status === STATUS_MISSED_DEADLINE || paymentDueDate.isBefore(moment())) {
```

Comment:
`isPassedCourseRun` will check for `STATUS_PASSED` or `STATUS_MISSED_DEADLINE`, and if it is `STATUS_PASSED` then the user would have paid and would be in the `if` part of the expression. So I think you can get rid of this `if` expression entirely since it's always true I think